fastpair/rust/demo/rust/src/decoder.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::fs;
+
+use serde::Deserialize;
+
+use bluetooth::ServiceData;
+
+use crate::advertisement::ModelId;
+
+/// Unit struct providing parsing operations for Fast Pair advertisements.
+pub(crate) struct FpDecoder;
+
+/// Holds Fast Pair device information parsed from JSON.
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct DeviceInfo {
+    image_url: String,
+    name: String,
+}
+
+/// Holds top-level Fast Pair information parsed from JSON. See `local`
+/// directory for format.
+#[derive(Deserialize)]
+struct JsonData {
+    device: DeviceInfo,
+}
+
+impl FpDecoder {
+    /// Retrieve device information for the provided Model ID. Currently,
+    /// this information is saved locally. In the future, this should instead
+    /// be retrieved from a remote server and cached.
+    /// b/294456411
+    pub(crate) fn get_device_info_from_model_id(
```

Comment:
This logic doesn't seem well coupled to decoding, so I'd make a trait for it, with a fs-backed struct implementing that trait. Also wouldn't hurt to make the root dir configurable instead of assuming `.`.